Packages/com.unity.inputsystem/InputSystem/Editor/AssetImporter/InputActionImporter.cs

```diff
@@ -256,11 +256,11 @@ private static void GenerateWrapperCode(AssetImportContext ctx, InputActionAsset
 
             if (InputActionCodeGenerator.GenerateWrapperCode(wrapperFilePath, asset, options))
             {
-                // When we generate the wrapper code cs file during asset import, we cannot call ImportAsset on that directly because
-                // script assets have to be imported before all other assets, and are not allowed to be added to the import queue during
-                // asset import. So instead we register a callback to trigger a delayed asset refresh which should then pick up the
-                // changed/added script, and trigger a new import.
-                EditorApplication.delayCall += AssetDatabase.Refresh;
```

Comment:
> use a source generator (Roslyn) for this kind of feature, but cannot be reliably setup in current Unity version. 

Why? And how is that handled in similar cases among the editor then?
It sounds like you put in quite an effort on researching the details around it, it would be interesting to hear more about it.

Reply:
I consulted the asset database team for the best way forward with this one. It seems its not recommended/support to generate c-sharp code assets from another assets. This is what we do.

Roslyn generators have been supported for some time https://docs.unity3d.com/6000.0/Documentation/Manual/roslyn-analyzers.html and it works well, I wrote a couple of generators myself to explore previously and its used by e.g. DOTS, but it doesn't support mapping assets to them. So they are usable for attributed code typically. Also UI Toolkit uses them I believe with the e.g. UxmlElement tags etc.

Regarding similar cases I am not aware of any other case where source files (assets) are generated from another asset. Are you aware of any?

Reply:
Thank you for elaborating.

That's weird, is the inputactionasset the only asset that is suspect to be imported then? Or how come this case is so unique? I guess what I am after is, why do we generate it from another asset, and why can't we solve this like the import of any other asset?

Reply:
I am just asking the five why's to understand the root of the problem.

Reply:
Importing isn't unique at all, but the usual import is just that we parse JSON and create ScriptableObjects (InputActionAsset). The problematic part is the writing .cs files when source generation option is checked since code needs to be imported before assets.

Reply:
Its not possible to customly import scripts

Reply:
Okay I think I get the gist of it, it sounds a bit like a devils cycle to write it, It's a bummer that there is no legitimate way to do that properly. But as long as it works it works I guess. Thanks for explaining.

Reply:
Will mark this resolved